platform/stm32f4xx/export/platform/spi_bus.hpp

```diff
@@ -461,57 +531,96 @@ void spi_bus< spi_config >::start_xfer()
     SPI_I2S_DMACmd(spi, SPI_I2S_DMAReq_Rx | SPI_I2S_DMAReq_Tx , ENABLE);
 }
 
-template< class spi_config >
-void spi_bus< spi_config >::irq_handler()
+template< class spi_i2s_config >
+void spi_i2s_bus< spi_i2s_config >::irq_handler()
 {
-    constexpr auto tx_dma   = dma::get_stream< spi_config::m_dma_tx_stream >();
-    constexpr auto rx_dma   = dma::get_stream< spi_config::m_dma_rx_stream >();
+    constexpr auto rx_irqn = dma::get_irqn< spi_i2s_config::m_dma_rx_stream >();
+    constexpr auto tx_irqn = dma::get_irqn< spi_i2s_config::m_dma_tx_stream >();
 
-    constexpr auto tx_tc_if = dma::get_tc_if< spi_config::m_dma_tx_stream >();
-    constexpr auto rx_tc_if = dma::get_tc_if< spi_config::m_dma_rx_stream >();
+    constexpr auto tx_dma   = dma::get_stream< spi_i2s_config::m_dma_tx_stream >();
+    constexpr auto rx_dma   = dma::get_stream< spi_i2s_config::m_dma_rx_stream >();
 
-    constexpr auto spi      = pick_spi();
+    constexpr auto tx_tc_if = dma::get_tc_if< spi_i2s_config::m_dma_tx_stream >();
+    constexpr auto rx_tc_if = dma::get_tc_if< spi_i2s_config::m_dma_rx_stream >();
 
+    constexpr auto tx_ht_if = dma::get_ht_if< spi_i2s_config::m_dma_tx_stream >();
+    constexpr auto rx_ht_if = dma::get_ht_if< spi_i2s_config::m_dma_rx_stream >();
+
+    constexpr auto spi = pick_spi();
 
     if (!(m_status & tx_complete)) {
         auto tx_tc = DMA_GetITStatus(tx_dma, tx_tc_if);
+        auto tx_ht = DMA_GetITStatus(tx_dma, tx_ht_if);
 
         if (tx_tc) {
             // Complete TX transaction
-            constexpr auto tx_tc_flag = dma::get_tc_flag< spi_config::m_dma_tx_stream >();
+            constexpr auto tx_tc_flag = dma::get_tc_flag< spi_i2s_config::m_dma_tx_stream >();
 
             if (!(m_status & tx_hidden)) {
                 m_event_handler(channel::tx, event::tc, m_tx_size);
             }
 
             DMA_ClearFlag(tx_dma, tx_tc_flag);
             DMA_ClearITPendingBit(tx_dma, tx_tc_if);
-            DMA_ITConfig(tx_dma, DMA_IT_TC, DISABLE);
+            DMA_ITConfig(tx_dma, DMA_IT_TC | DMA_IT_HT, DISABLE);
 
             m_status |= tx_complete;
+        } else if (tx_ht) {
+            constexpr auto tx_ht_flag = dma::get_ht_flag< spi_i2s_config::m_dma_tx_stream >();
+
+            uint32_t tx_left = DMA_GetCurrDataCounter(tx_dma);
+
+            if (spi_i2s_config::m_interface == spi_i2s_interface::i2s) {
+                tx_left *= 2;
+            }
+
+            if (!(m_status & tx_hidden)) {
+                m_event_handler(channel::tx, event::ht, m_tx_size - tx_left);
+            }
+
+            DMA_ClearFlag(tx_dma, tx_ht_flag);
+            DMA_ClearITPendingBit(tx_dma, tx_ht_if);
+
+            irq_manager::clear(tx_irqn);
+            irq_manager::unmask(tx_irqn);
         }
     }
 
-    auto rx_tc = DMA_GetITStatus(rx_dma, rx_tc_if);
 
     if (!(m_status & rx_complete)) {
+        auto rx_tc = DMA_GetITStatus(rx_dma, rx_tc_if);
+        auto rx_ht = DMA_GetITStatus(tx_dma, rx_ht_if);
```

Comment:
Let's make a bit of optimization. Move DMA_GetITStatus(...) inside if statements. In such way only call will be executed when handling rx_tc event